src/EventEmitter.ts

```diff
@@ -0,0 +1,177 @@
+/**
+ * A generic handler for incoming event data.
+ */
+type THandler<T> = (data: T) => void;
+
+/**
+ * Simple object for keeping track of event listeners.
+ */
+interface Events<T> {
+  [index: string]: Array<THandler<T>>;
+}
+
+const findIndex =
+  [].findIndex ||
+  // IE11 support
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex#Polyfill
+  function(
+    predicate: (value: never, index: number, obj: never[]) => boolean,
+    thisArg?: any
+  ) {
+    if (this == null) {
+      throw new TypeError('"this" is null or not defined');
+    }
+
+    const o = Object(this);
+    // tslint:disable-next-line
+    const len = o.length >>> 0;
+
+    if (typeof predicate !== 'function') {
+      throw new TypeError('predicate must be a function');
+    }
+
+    let k = 0;
+    while (k < len) {
+      const kValue = o[k];
+      if (predicate.call(thisArg, kValue, k, o)) {
+        return k;
+      }
+
+      k++;
+    }
+    return -1;
+  };
+
+/**
+ * Used to determine equivalancy of two handlers.
+ */
+function isRegistered(value: THandler<any>) {
+  return value === this;
+}
+
+/**
+ * Exposes a limited subset of event emitter
+ * functionality to ensure external modules can
+ * not self-emit.
+ */
+export class ExposedEventEmitter<T> {
```

Comment:
Is there is a better way to make it so JS people can't access dispatch?

Reply:
Sadly, I don't know of a a better option. The delegation would be more concise in vanilla JS, but the structure would be pretty much the same.